pkg/operator/controllers/checkers/serviceprincipalchecker/checker.go

```diff
@@ -0,0 +1,61 @@
+package serviceprincipalchecker
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/Azure/ARO-RP/pkg/api/validate/dynamic"
+	"github.com/Azure/ARO-RP/pkg/util/aad"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	"github.com/Azure/ARO-RP/pkg/util/clusterauthorizer"
+)
+
+type servicePrincipalChecker interface {
+	Check(ctx context.Context, AZEnvironment string) error
+}
+
+type checker struct {
+	log *logrus.Entry
+
+	credentialsGetter      func(ctx context.Context) (*clusterauthorizer.Credentials, error)
+	spValidatorConstructor func(azEnv *azureclient.AROEnvironment) (dynamic.ServicePrincipalValidator, error)
+}
+
+func newServicePrincipalChecker(log *logrus.Entry, kubernetescli kubernetes.Interface) servicePrincipalChecker {
```

Comment:
we probably don't need to return the interface type here. We are losing access to the struct fields because of that and will need to typecast it. Returning the actual type will work just as well

Reply:
But we don't need access to the fields. We only need access to the method defined in the interface.

* When testing the `checker` struct itself - we have access to the fields.
* When testing `Reconciler` - we do not have access to the feilds and we replace the whole thing by a mock which statisfies `servicePrincipalChecker` interface.

Reply:
> but this is just a bad pattern though

I don't agree. Ideally no one should be concerned about underlying implementation inside of the package hence we should keep stuff non-exported untill there is a really good reason for this.

> just the other day I had to fix a bunch of frontend typecasting because NewFrontend didn't return a frontend struct but an interface type.

I would like to look into this with you. You are probably talking about our front-end unit tests which are more like integration tests. Yes, we have to cast into a specific type there, but it is not the issue with the pattern. It is an issue with the tests. Hapyp to discuss this.